lib/_stream_writable.js

```diff
@@ -173,25 +171,17 @@ function WritableState(options, stream, isDuplex) {
   // should return false. This is needed since when autoDestroy
   // is disabled we need a way to tell whether the stream has failed.
   this.errored = false;
+}
 
-  // Count buffered requests
-  this.bufferedRequestCount = 0;
-
-  // Allocate the first CorkedRequest, there is always
-  // one allocated and free to use, and we maintain at most two
-  const corkReq = { next: null, entry: null, finish: undefined };
-  corkReq.finish = onCorkedFinish.bind(undefined, corkReq, this);
-  this.corkedRequestsFree = corkReq;
+function resetBuffer(state) {
+  state.buffered = [];
+  state.bufferedIndex = 0;
```

Comment:
Would it not be possible to shift or pop entries of the buffered array and then to rely upon `buffered.length`? That would probably simplify the logic even further?

Reply:
we could `shift()` but that will be significantly slower. The array "left trimming" optimization is no longer as fast in V8.

Reply:
> The array "left trimming" optimization is no longer as fast in V8.

I know there used to be an upper bound for the array size that it works for but otherwise it pretty much stayed linear (V8 internally using a start pointer). Did you run the benchmarks against a shift implementation as well and do you know when it regressed?

Reply:
```
# bufferedIndex
streams/writable-manywrites.js callback='no' writev='no' sync='no' n=1000000           ***      3.07 %       ±0.88% ±1.17% ±1.50%
streams/writable-manywrites.js callback='no' writev='no' sync='yes' n=1000000          ***     -3.57 %       ±1.77% ±2.36% ±3.07%
streams/writable-manywrites.js callback='no' writev='yes' sync='no' n=1000000          ***     22.04 %       ±1.33% ±1.77% ±2.29%
streams/writable-manywrites.js callback='no' writev='yes' sync='yes' n=1000000                 -3.01 %       ±4.03% ±5.38% ±7.03%
streams/writable-manywrites.js callback='yes' writev='no' sync='no' n=1000000                   1.84 %       ±1.98% ±2.62% ±3.39%
streams/writable-manywrites.js callback='yes' writev='no' sync='yes' n=1000000          **     -3.24 %       ±2.22% ±2.95% ±3.85%
streams/writable-manywrites.js callback='yes' writev='yes' sync='no' n=1000000         ***      8.58 %       ±1.38% ±1.83% ±2.36%
streams/writable-manywrites.js callback='yes' writev='yes' sync='yes' n=1000000         **     -2.38 %       ±1.60% ±2.14% ±2.81%

# shift()
streams/writable-manywrites.js callback='no' writev='no' sync='no' n=1000000           ***     -4.58 %       ±1.00% ±1.33% ±1.72%
streams/writable-manywrites.js callback='no' writev='no' sync='yes' n=1000000          ***     -3.62 %       ±1.96% ±2.61% ±3.41%
streams/writable-manywrites.js callback='no' writev='yes' sync='no' n=1000000          ***     22.19 %       ±1.23% ±1.62% ±2.10%
streams/writable-manywrites.js callback='no' writev='yes' sync='yes' n=1000000         ***     -5.18 %       ±2.27% ±3.04% ±4.02%
streams/writable-manywrites.js callback='yes' writev='no' sync='no' n=1000000          ***     -5.23 %       ±1.15% ±1.53% ±1.98%
streams/writable-manywrites.js callback='yes' writev='no' sync='yes' n=1000000          **     -2.56 %       ±1.75% ±2.34% ±3.08%
streams/writable-manywrites.js callback='yes' writev='yes' sync='no' n=1000000         ***      8.81 %       ±2.88% ±3.82% ±4.96%
streams/writable-manywrites.js callback='yes' writev='yes' sync='yes' n=1000000        ***     -4.78 %       ±1.82% ±2.45% ±3.22%
```

Reply:
I personally don't mind the performance impact but I don't think that matches consensus.

Reply:
> do you know when it regressed?

No idea.

Reply:
The async case is around 5-7% slower (with high margin of error).